[AUTH-133] Add pod security admission control docs (#47647)
openshift/openshift-docs: modules/security-context-constraints-psa-synchronization.adoc

// Module included in the following assemblies:
//
// * authentication/managing-security-context-constraints.adoc

> :_content-type: CONCEPT

Reviewer: Maybe Pod Security Admission should have its own new page that would elaborate on global configuration, label synchronization and alerts?

Author: @stlaz There are a couple of things to think through, here:

1. Currently, this is all included at the end of `managing-security-context-constraints.adoc`. I'm not opposed to moving new PSA content into a document that exists at the same level as that one, though.
2. Which information that is _required_ for this feature to GA is currently missing? Which information that would be nice to have, but is not critical for GA is missing and could be bug fixed or enhanced later?

Reviewer: 1. I think this deserves its own place in the docs
2. I don't think there's missing information

Author: 1. I can try that out today. 
2. These are just:
2a. Label `security.openshift.io/scc.podSecurityLabelSync=false`?
2b.  Not sure. All namespaces that aren't prefixed w/`openshift-`?

Author: Actually, 2a should be covered by `modules/security-context-constraints-psa-opting.adoc` at the moment. Do you concur?

Reviewer: Ah, true. We are still missing the information that the label sync is disabled on user-created "openshift-"-prefixed namespaces but can be still opted in there, and why.

We are also missing the information that the namespaces that are deployed by the installer cannot be opted in to the sync.

Author: Updated.

Reviewer: Where is this new documentation, then?

Author: @stlaz Actually, don't have this:

> label sync is disabled on user-created "openshift-"-prefixed namespaces but **can be still opted in there, and why.**

I assume the answer to the first part is `security.openshift.io/scc.podSecurityLabelSync=true`, but I'm not sure about "why."

Reviewer: The "why" is because those are considered system namespaces and as such, the expectation is that there exists a controller that takes care of them.

The "why" on the installer payload namespaces is because these namespaces are maintained by OpenShift developers.

[id="security-context-constraints-psa-synchronization_{context}"]
= Security context constraint synchronization with pod security standards

> In addition to the global pod security admission control configuration, {product-title} translates pod security profiles that use the `warn` and `audit` labels for every namespace.

Reviewer: I don't understand this sentence. What does it translate into what?

The profiles are not using the labels. The labels mark which profiles should be considered during admission.

Author: So:

```suggestion
{product-title} applies labels to namespaces to identify which pod security profiles to consider during admission. 
```

?

Reviewer: `In addition to the global pod security admission control configuration, {product-title} applies Pod Security Admission Control warn and audit labels to namespaces according to the SCC permissions of the Service Accounts in the given namespace.`

Author: ```suggestion
In addition to the global pod security admission control configuration, {product-title} applies pod security admission control `warn` and `audit` labels to namespaces according to the SCC permissions of the service accounts that are in a given namespace.
```

Reviewer: :+1:

> By using a controller, clusters examine `ServiceAccount` object permissions to apply security context constraints to each namespace. Security context constraints (SCCs) are mapped to pod security profiles based on their field values. Pod security admission control `warn` and `audit` labels are applied to the most privileged pod security profile in the namespace to prevent warnings and audit logging as pods are created. 

Reviewer: Security Context Constraints are not applied to namespaces, their translation into a Pod Security profile is.

Author: ```suggestion
By using a controller, clusters examine `ServiceAccount` object permissions to apply pod security profiles constraints to each namespace. Security context constraints (SCCs) are mapped to pod security profiles based on their field values; the controller uses these translated profiles. Pod security admission control `warn` and `audit` labels are applied to the most privileged pod security profile in the namespace to prevent warnings and audit logging as pods are created. 
```

> This controller cannot use automatic namespace labels for pods that are created without an intermediary, like the Deployment controller. This restriction prevents the SCC privileges of a user from being used to create a pod. If possible, do not run pods directly.

Reviewer: We should probably rephrase this, the labels aren't applied or "used" for pods, they are applied to namespaces based on a consideration of the namespace-local SA privileges.

"Direct pod application" without an intermediary "pod controller" cannot be considered.

```
This restriction prevents the SCC privileges of a user from being used to create a pod.
```
This is rather inaccurate, they are still used, only if they appear more permissive than the pod security admission enforcement level, the pod will still not be admitted. On the other hand, if the permissions are less permissive, they will be very much applied.

Author: ```suggestion
The controller applies labels to namespaces based on consideration of namespace-local service account privileges.

Direct pod application without an intermediary pod controller cannot be considered.  If possible, do not run pods directly.

The SCC privileges of a user are only used for pod creation if they are less permissive than the pod security admission enforcement level. If those privileges are less permissive, they are applied.

```

So, something like this? @stlaz

Reviewer: Just leave out all of
```
Direct pod application without an intermediary pod controller cannot be considered. 

If possible, do not run pods directly. The SCC privileges of a user are only used for pod creation if they are less permissive than the pod security admission enforcement level. If those privileges are less permissive, they are applied.
```
Maybe in favour of something like:
```
Note that applying pod directly might leverage SCC privileges of the user running the pod, however user privileges are not considered during this automatic labelling.
```

Reviewer: :+1:

// Module included in the following assemblies:
//
// * authentication/managing-security-context-constraints.adoc

:_content-type: CONCEPT
[id="security-context-constraints-psa-synchronization_{context}"]
= Security context constraint synchronization with pod security standards

> {product-title} implements Kubernetes pod security admission control. Globally, the `privileged` profile is enforced, and the `restricted` profile is used for warnings and audits.

Author: Sure. 

Lowercase aside from proper branding or API object contexts.

> {product-title} implements Kubernetes pod security admission control. Globally, the `privileged` profile is enforced, and the `restricted` profile is used for warnings and audits.

Author: ```suggestion
{product-title} implements Kubernetes pod security admission. Globally, the `privileged` profile is enforced, and the `restricted` profile is used for warnings and audits.
```

In addition to the global pod security admission control configuration, a pod admission controller applies pod security admission control `warn` and `audit` labels to namespaces according to the SCC permissions of the service accounts that are in a given namespace.

> The controller also examines `ServiceAccount` object permissions to apply pod security profiles constraints to each namespace. Security context constraints (SCCs) are mapped to pod security profiles based on their field values; the controller uses these translated profiles. Pod security admission control `warn` and `audit` labels are set to the most privileged pod security profile found in the namespace to prevent warnings and audit logging as pods are created.

Author: ```suggestion
The controller examines `ServiceAccount` object permissions to apply pod security profiles constraints to each namespace. Security context constraints (SCCs) are mapped to pod security profiles based on their field values; the controller uses these translated profiles. Pod security admission `warn` and `audit` labels are set to the most privileged pod security profile found in the namespace to prevent warnings and audit logging as pods are created.
```

> The controller applies labels to namespaces based on consideration of namespace-local service account privileges.

Author: Still necessary?

Applying pods directly might use the SCC privileges of the user who runs the pod. However, user privileges are not considered during automatic labeling.

[IMPORTANT]
====

> SCC synchronization is disabled for user-created namespaces that are named with the prefix `openshift-`.

Author: We cannot link to pages in our docs set from within modules. It's a frustrating limitation. There is a direction in the note, at least. :\

> SCC synchronization is disabled for user-created namespaces that are named with the prefix `openshift-`.
> 
> Namespaces that the installer deploys cannot be opted into SCC synchronization.

Author: @stlaz 

```suggestion
SCC synchronization is disabled for user-created namespaces that are named with the prefix `openshift-`. Namespaces that have this prefix are typically system namespaces; by convention, a controller exists to manage them. 

You can enable SCC synchronization in namespaces that have the prefix `openshift-` by setting the value of the `security.openshift.io/scc.podSecurityLabelSync` label to `false`. 

Namespaces that the installer deploys cannot be opted into SCC synchronization. These namespaces are system namespaces.
```

In addition to the global pod security admission control configuration, a pod admission controller applies pod security admission control `warn` and `audit` labels to namespaces according to the SCC permissions of the service accounts that are in a given namespace.

The controller examines `ServiceAccount` object permissions to apply pod security profiles constraints to each namespace. Security context constraints (SCCs) are mapped to pod security profiles based on their field values; the controller uses these translated profiles. Pod security admission `warn` and `audit` labels are set to the most privileged pod security profile found in the namespace to prevent warnings and audit logging as pods are created.

Namespace labeling is based on consideration of namespace-local service account privileges.

Applying pods directly might use the SCC privileges of the user who runs the pod. However, user privileges are not considered during automatic labeling.

[IMPORTANT]
====

> SCC synchronization is disabled by default for user-created namespaces that are named with the prefix `openshift-`. Namespaces that have this prefix are typically system namespaces; by convention, a controller exists to manage them. 

Reviewer 2: ```suggestion
SCC synchronization is disabled by default for namespaces that are named with the prefix `openshift-`. Namespaces that have this prefix are typically system namespaces; by convention, a controller exists to manage them. 
```

What do you mean by "user-created"? I understand that if the namespace is prefixed with `openshift` then they will not be synced by default.

Reviewer: User-created means not created by OpenShift.

TODO (me): propose an improvement to this part of the docs

Reviewer: > Also, we need to clarify here that those are not enforced as restricted for now but will be in the next release.

Perhaps in the release notes, not here.

Reviewer: > What do you mean by "user-created"? I understand that if the namespace is prefixed with openshift then they will not be synced by default.

```
Namespaces with an "openshift-" name prefix that were not created by the system during the installation are not being synchronized by default.
```
Maybe something like this?

> You can enable SCC synchronization in namespaces that have the prefix `openshift-` by setting the value of the `security.openshift.io/scc.podSecurityLabelSync` label to `false`. 

Reviewer 2: ```suggestion
You can enable SCC synchronization in namespaces that have the prefix `openshift-` by setting the value of the `security.openshift.io/scc.podSecurityLabelSync` label to `true`. 
```

Reviewer: yes please! :+1:

> Namespaces that the installer deploys cannot be opted into SCC synchronization. These namespaces are system namespaces.

Reviewer 2: We need to describe exactly what are the namespaces. That is not 100% true ( i.e openshift-operators is created by the installed and is not in this list. ) 
```suggestion
Namespaces which are considered as part of OCP payload do not allow SCC synchronization. You can verify the [list](https://github.com/openshift/cluster-policy-controller/blob/release-4.11/pkg/psalabelsyncer/podsecurity_label_sync_controller.go#L42-L111) of namespaces those is not allowed the synchronization.
```

Reviewer: indeed, as I stated in my comments about this being vague on purpose and openshift-operators NS being an exception

// Module included in the following assemblies:
//
// * authentication/managing-security-context-constraints.adoc

:_content-type: CONCEPT
[id="security-context-constraints-psa-synchronization_{context}"]
= Security context constraint synchronization with pod security standards

> {product-title} ships link:https://kubernetes.io/docs/concepts/security/pod-security-admission[Kubernetes pod security admission]. Globally, the `privileged` profile is enforced, and the `restricted` profile is used for warnings and audits.

Author: ```suggestion
{product-title} includes link:https://kubernetes.io/docs/concepts/security/pod-security-admission[Kubernetes pod security admission]. Globally, the `privileged` profile is enforced, and the `restricted` profile is used for warnings and audits.
```

> {product-title} implements link:https://kubernetes.io/docs/concepts/security/pod-security-admission[Kubernetes pod security admission]. Globally, the `privileged` profile is enforced, and the `restricted` profile is used for warnings and audits.

Reviewer: ```suggestion
{product-title} ships link:https://kubernetes.io/docs/concepts/security/pod-security-admission[Kubernetes pod security admission]. Globally, the `privileged` profile is enforced, and the `restricted` profile is used for warnings and audits.
```

Author: We'll probably end up arriving at a different word during peer review, but :+1: for now.

> In addition to the global pod security admission control configuration, a pod admission controller applies pod security admission control `warn` and `audit` labels to namespaces according to the SCC permissions of the service accounts that are in a given namespace.

Reviewer: ```suggestion
In addition to the global pod security admission control configuration, a controller exists that applies pod security admission control `warn` and `audit` labels to namespaces according to the SCC permissions of the service accounts that are in a given namespace.
```

> The controller examines `ServiceAccount` object permissions to apply pod security profiles constraints to each namespace. Security context constraints (SCCs) are mapped to pod security profiles based on their field values; the controller uses these translated profiles. Pod security admission `warn` and `audit` labels are set to the most privileged pod security profile found in the namespace to prevent warnings and audit logging as pods are created.

Reviewer: ```suggestion
The controller examines `ServiceAccount` object permissions to use security context constraints in each namespace. Security context constraints (SCCs) are mapped to pod security profiles based on their field values; the controller uses these translated profiles. Pod security admission `warn` and `audit` labels are set to the most privileged pod security profile found in the namespace to prevent warnings and audit logging as pods are created.
```

Namespace labeling is based on consideration of namespace-local service account privileges.

Applying pods directly might use the SCC privileges of the user who runs the pod. However, user privileges are not considered during automatic labeling.

[IMPORTANT]
====
Namespaces that have an "openshift-" name prefix and were not created by the system during the installation are not synchronized by default. 

> Namespaces that have the "openshift-" prefix are typically system namespaces; by convention, a controller exists to manage them. 

Reviewer: ```suggestion
Namespaces that have the "openshift-" prefix are typically system namespaces; by convention, a controller should exist to manage them. 
```

In addition to the global pod security admission control configuration, {product-title} translates pod security profiles that use the `warn` and `audit` labels for every namespace.

> By using a controller, clusters examine `ServiceAccount` object permissions to apply pod security profiles constraints to each namespace. Security context constraints (SCCs) are mapped to pod security profiles based on their field values; the controller uses these translated profiles. Pod security admission control `warn` and `audit` labels are applied to the most privileged pod security profile in the namespace to prevent warnings and audit logging as pods are created. 

Reviewer: What's a pod security profile constraint?

Author: ```suggestion
By using a controller, clusters examine `ServiceAccount` object permissions to apply pod security profiles to each namespace. Security context constraints (SCCs) are mapped to pod security profiles based on their field values; the controller uses these translated profiles. Pod security admission control `warn` and `audit` labels are applied to the most privileged pod security profile in the namespace to prevent warnings and audit logging as pods are created. 
```

Reviewer: ```
pod security admission control `warn` and `audit` labels are set to the most privileged pod security profile found in the namespace to prevent warnings and audit logging as pods are created.
```
